File: test-suite/src/test/java/io/micronaut/kafka/docs/consumer/errors/RetryTopicProductDltListener.java
package io.micronaut.kafka.docs.consumer.errors;

import io.micronaut.configuration.kafka.annotation.KafkaListener;
import io.micronaut.configuration.kafka.annotation.Topic;
import io.micronaut.context.annotation.Requires;
import io.micronaut.messaging.MessageHeaders;

@Requires(property = "spec.name", value = "RetryTopicProductListenerTest")
// tag::dlt[]
@KafkaListener("product-retry-dlt-group")
public class RetryTopicProductDltListener {

    @Topic("products-dlt")
    void receive(String product, MessageHeaders headers) {
        String originalTopic = headers.get("micronaut-kafka-original-topic", String.class).orElse("unknown");
        System.out.printf("Routing %s from %s to the dead letter topic%n", product, originalTopic);
    }
}
// end::dlt[]

File: test-suite-kotlin/src/test/kotlin/io/micronaut/kafka/docs/consumer/errors/RetryTopicProductDltListener.kt
package io.micronaut.kafka.docs.consumer.errors

import io.micronaut.configuration.kafka.annotation.KafkaListener
import io.micronaut.configuration.kafka.annotation.Topic
import io.micronaut.context.annotation.Requires
import io.micronaut.messaging.MessageHeaders

@Requires(property = "spec.name", value = "RetryTopicProductListenerTest")
// tag::dlt[]
@KafkaListener("product-retry-dlt-group")
class RetryTopicProductDltListener {

    @Topic("products-dlt")
    fun receive(product: String, headers: MessageHeaders) {
        val originalTopic = headers.get("micronaut-kafka-original-topic", String::class.java).orElse("unknown")
        println("Routing $product from $originalTopic to the dead letter topic")
    }
}
// end::dlt[]
